### ISSUE TITLE
twisted.internet.task.deferLater()

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1875]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)| @mithrandi reported|
|-|-|
|Trac ID|trac#1875|
|Type|enhancement|
|Created|2006-06-28 23:45:49Z|
|Branch|https://github.com/twisted/twisted/tree/task-deferlater-1875-2|

See attached patch.

Attachments:

* [deferlater.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f0d/f0d579502820ee87140d20364532ce6a63a5ae33/46c8d22363e8cf45166df04a9ff6d14917b287dd.diff) (3286 bytes) - added by mithrandi on 2006-06-29 00:00:04Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1875 1875
type__enhancement enhancement
reporter__mithrandi mithrandi
priority__highest highest
milestone__ 
branch__branches_task_deferlater_1875_2 branches/task-deferlater-1875-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1151538349000000 1151538349000000
changetime__1203040953000000 1203040953000000
version__None None
owner__ 
cc__jknight cc__exarkun cc__radix cc__jerub cc__mithrandi cc__strank cc__therve cc__yaubi
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)|@mithrandi set owner to @jerub|
|-|-|

Attaching updated patch.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @mithrandi|
|-|-|

Okay, I've had a good look at the patch, and I'm satisfied that it does what it says it does. I have two concerns.

One: I want to run this past glyph or exarkun, just to make sure that this code is _worth_ going into Twisted. It looks very similar to reactor.callLater, I feel there should be one obvious way of doing things. Should we deprecate callLater and point to deferLater? Or should we put guidelines in here for where deferLater should be used instead of callLater.

Two: ```def _callLater(self, *a, **kw):``` should probably be written: ```def __callLater(self, delay, func):```. The function is private, and is only invoked like this: ```self._callLater(delay, lambda: d.callback(None))```.

Please update the patch with the above modifications, and I'll take responsibility for merging the updated patch if/when I get approval/guidance.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub commented|
|-|-|

The double underscore in front of ```__callLater``` in my comment is a typo. It should be a single underscore.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub commented|
|-|-|

Branch created: task-deferlater-1875

I've altered the twisted.web2.test.test_http tests to import deferLater, and use that instead. This has managed to reduce the amount of code required, which is highly satisfactory.

More tests for deferLater would be good - we don't currently have a test for if the function raises an error and fires an errback.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

I'm confused, why does it do an addCallback for you? That seems unnecessary and redundant. I'd suggest using the function as it was in test_http. Simple, clean, no extraneous gunk.

def deferLater(secs):
    d = defer.Deferred()
    reactor.callLater(secs, d.callback, None)
    return d

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_7"></a>|@radix commented|
|-|-|

What James said.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Why is

```
deferLater(n).addCallback(lambda ign: f(x))
```

preferable to

```
deferLater(n, f, x)
```

?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

How about a compromise?  The function can default to ```lambda: None```, then both of the above will work.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_10"></a>|@jyknight commented|
|-|-|

It's preferable for simplicity of API. The deferred is the primary concept here, and it has a perfectly good API already which there's no good reason to hide.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

It's not _perfectly_ good.  The feature it is missing in this case is invoking a callback without passing any arguments.  This is what is addressed by allowing a function and arguments to be passed directly to deferLater.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_12"></a>|@jyknight commented|
|-|-|

Same thing is true for all deferred callback functions. This shouldn't be special.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

But there's never any possibility of a useful value being passed in here.  The fact that the call is being made is the entirety of the information being delivered.

**Comment by twisted-trac** [comment]
|<img alt="strank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|strank commented|
|-|-|

On the mailing list today, I asked about a way to get the result of a task scheduled with callLater. deferLater nearly solves that problem.

As for the API, it would be easy to allow both variants: simply change the callable argument to default to None::

```
    def __call__(self, delay, callable_=None, *args, **kw):
        d = defer.Deferred() 
        if callable_ is not None:
            d.addCallback(lambda dummy: callable(*args, **kw)) 
        self._callLater(delay, lambda: d.callback(None)) 
        return d 
```

If there has to be only one way to call it, I think that ```deferLater(n, f, x)``` is much better, for the same reasons as exarkun mentioned.

What I am missing is a way to get to the ```DelayedCall``` instance returned by ```callLater```.
I would simply add a task attribute to the returned Deferred::

```
        ...
        d.task = self._callLater(delay, lambda: d.callback(None))
        return d 
```

but I am not sure if this is nice enough.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve commented|
|-|-|

Can we reach an agreement on this? We can probably have the both interfaces, one basic and another a little more high-level?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve commented|
|-|-|

ping!

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)|@mithrandi removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

Hmm, I'm thinking about the issue raised by strank now.  What if you want to cancel the deferLater's DelayedCall?

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_20"></a>|@jyknight commented|
|-|-|

Hmm, I think I heard something about adding cancelable Deferreds at some point. Of course, that still doesn't let you reschedule/etc, but that seems okay to me: the more verbose API is still available for such special uses.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_21"></a>|@jyknight commented|
|-|-|

If it helps, I'll just state I don't care anymore which of the API alternatives is added, they're really both fine. Just, please don't add both or a hybrid, in some attempt to please me.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

(In [22507]) Branching to 'task-deferlater-1875-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

Okay

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

Missed a thing...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

Missed thing was r22515.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_26"></a>|@therve set owner to @exarkun|
|-|-|

 * deferLater doesn't document its clock argument
 * in writing-standard.xhtml, the example is wrong because it doesn't pass the reactor as argument

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun removed owner|
|-|-|

Doh, I messed that up pretty good. :)

Fixed those things.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_28"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun set status to `closed`|
|-|-|

(In [22548]) Merge task-deferlater-1875-2

Author: mithrandi, jerub, exarkun
Reviewer: therve
Fixes [#1875](https://github.com/twisted/twisted/issues/1875)

Add twisted.internet.task.deferLater, an API for scheduling a future call
and getting a Deferred which is called back with its result.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
